### PR TITLE
Add support for custom player relative placeholders

### PR DESCRIPTION
**Pull request by kezz**
This PR extends the existing API to include support for player relative placeholders. This works with no modification of the existing API to ensure backwards compatibility.

It works by creating a new type of `PlaceholderReplacer` that provides a reference to the `Player` that is being sent the hologram. The only code that is edited is the code in the old and new `ProtocolLibHookImpl` classes. This means that player relative placeholders just won't work if the user doesn't have ProtocolLib (just like the existing ones).

Basically while making this I aimed to make it not look like a new feature, but rather an extension of the existing player relative placeholders (although the code has a fair few additions!)

To begin with I've just added a `{gamemode}` placeholder but the api is exactly the same as the current one so it won't be hard at all to add more.

### CODE REVIEW

**Comment by Kezz101** [comment]
One quick note, I haven't been able to test this as I don't have enough free time at the moment to compile everything with these blasted versioning errors 😢 

Would it be possible for someone who has this set up for easy compilation to make this into a jar and I'll do some thorough testing!

Alternatively, give it a few days and I'll probably have mustered up the courage to compile it myself!

**Comment by FlashOnFire** [comment]
These modifications require some changes to be compatible now, I made these changes and I compiled the plugin yesterday. I'll test all the modifications today.

**Comment by FlashOnFire** [comment]
doesn't work :/

**Comment by FlashOnFire** [comment]
It looks like if the placeholders aren't updated, They are simply replaced with empty strings...

**Comment by Niall7459** [comment]
https://www.spigotmc.org/resources/holographic-extension.18461/

**Comment by filoghost** [comment]
Thanks for the PR. Placeholders need a proper rework, and I want to do it in v3. However, I'm still concerned about performance, and that's why I didn't accept the PR before (sorry if I didn't answer). Per-player placeholders can be called many times a second (for example, using the &u animated rainbow color).

I'm trying to think a solution where plugins with slow/inefficient placeholders are displayed in timings in their respective section.

**Comment by filoghost** [comment]
Holographic Displays v3 will use packets only, and will support relative placeholders without ProtocolLib. As for performance, I will later add metrics to detect slow placeholders and show warnings on the console, but I will probably not be able to display something useful in timings.

I'm closing this as the implementation is completely different.
